File: printer_server/drivers/visitech/visitech_snip.py
from printer_server.extensions import socketio
from printer_server.hardware_configuration import driver_handles

visitech = driver_handles.visitech


@socketio.on("visitech_stop", namespace="/manual")
def visitechStop():
    """Turn off the LED in the light engine."""
    visitech.stop_sequencer()
    socketio.emit("visitech_update_led_state", False, namespace="/manual")
    socketio.emit("visitech_done", namespace="/manual")


@socketio.on("visitech_start", namespace="/manual")
def visitechProject(message):
    """Project the image with the given settings."""
    ledPower = int(message["ledPower"])
    repeat = int(message["repeat"])
    exposure = int(message["exposure"])
    led = int(message.get("led", 0))
    socketio.emit("visitech_update_led_state", True, namespace="/manual")
    visitech.project(exposure, ledPower, repeat, led_num=led)
    if repeat != 0:
        socketio.emit("visitech_update_led_state", False, namespace="/manual")
    socketio.emit("visitech_done", namespace="/manual")


@socketio.on("visitech_get_status", namespace="/manual")
def visitechStatus():
    socketio.emit(
        "visitech_return_status",
        visitech.read_all_status(warn="ALL"),
        namespace="/manual"
    )


def getLedStatus():
    return visitech.led_on

File: printer_server/drivers/visitech/visitech_snip.js
var enable_visitech_start_button = function () {
    $('#visitech-start-btn').prop('disabled', false);
    $('#visitech-start-btn').addClass('btn-outline-info');
    $('#visitech-start-btn').removeClass('btn-outline-secondary');
}

var disable_visitech_start_button = function () {
    $('#visitech-start-btn').prop('disabled', true);
    $('#visitech-start-btn').removeClass('btn-outline-info');
    $('#visitech-start-btn').addClass('btn-outline-secondary');
}

$(document).ready(function () {
    // Handles to user inputs
    var VisitechLedPowerSliderElement = document.getElementById("visitech-led-power-slider");
    var VisitechLedPowerLabelElement = document.getElementById("visitech-led-power-label");
    var VisitechExposureElement = document.getElementById("visitech-exposure-txt");

    // Set initial LED power slider label value
    VisitechLedPowerLabelElement.innerHTML = VisitechLedPowerSliderElement.value; // Display the default slider value

    // Update the VisitechLedPowerLabelElement with the current slider value
    VisitechLedPowerSliderElement.oninput = function () {
        VisitechLedPowerLabelElement.innerHTML = this.value;
    }

    // Light engine control stop button click function
    $("#visitech-stop-btn").click(function () {
        socket.emit("visitech_stop");
    });

    // Exposure text input change function
    $('#visitech-exposure-txt').on('change', function () {
        exposure = VisitechExposureElement.value;

        // Validate user input. Only allows positive integers > 0
        if (/^\d+$/.test(exposure) && exposure > 0) {
            VisitechExposureElement.classList.remove("is-invalid")
            enable_visitech_start_button();
        } else {
            VisitechExposureElement.classList.add("is-invalid")
            disable_visitech_start_button();
        }
    })

    $("#visitech-repeat-chkbx").click(function () {
        var repeatCheckboxElement = document.getElementById("visitech-repeat-chkbx");
        var repeat = Number(!repeatCheckboxElement.checked);
        if (repeat == 0) {
            VisitechExposureElement.classList.remove("is-invalid")
            $('#visitech-exposure-txt').prop('disabled', true);
        } else {
            $('#visitech-exposure-txt').prop('disabled', false);
        }
    });

    // Light engine control start button click function
    $("#visitech-start-btn").click(function () {
        var repeatCheckboxElement = document.getElementById("visitech-repeat-chkbx");
        var exposure = VisitechExposureElement.value;
        var repeat = Number(!repeatCheckboxElement.checked);
        var ledPower = VisitechLedPowerSliderElement.value;

        if (!/^\d+$/.test(exposure) && !exposure > 0) {
            if (repeat == 1) {
                VisitechExposureElement.classList.add("is-invalid")
                return
            } else {
                exposure = 1
            }
        }

        var led_1_element = document.getElementById(`led_1`);
        var led_1_checked = false;
        if (led_1_element != null) {
            led_1_checked = led_1_element.classList.contains("active");
        }

        console.log({ "repeat": repeat, "exposure": exposure, "ledPower": ledPower, "led": led_1_checked })
        socket.emit("visitech_start", { "repeat": repeat, "exposure": exposure, "ledPower": ledPower, "led": led_1_checked });
    });

    socket.on("update_visitech_led_status", function (message) {
        var statusElement = document.getElementById("visitech-status");
        if (message == true) {
            statusElement.classList.remove("invisible")
        } else {
            statusElement.classList.add("invisible")
        }
        $('#visitech-status').prop('disabled', message);
    });

});